internal/provider/resource_vault_cluster.go

```diff
@@ -610,6 +660,18 @@ func resourceVaultClusterUpdate(ctx context.Context, d *schema.ResourceData, met
 		return diagErr
 	}
 
+	// get plugins for plugin-name validation in getPluginConfig
+	plugins, err := clients.ListPlugins(ctx, client, loc, clusterID)
+	if err != nil {
+		log.Printf("[ERROR] Vault cluster (%s) failed to list plugins", clusterID)
+		return diag.FromErr(err)
+	}
+
+	newPluginConfig, diagErr := getPluginConfig(d, plugins.Plugins)
```

Comment:
I was a little unclear in start of why we are passing listed cluster plugins here to ensure that new plugin is part of cluster plugin. But then realized that listed cluster plugin can have all valid plugins which may be not registered by the cluster too (the response isRegistered will tell you which are part of the cluster yet). So this way you ensure that the new plugin is at least a right plugin name. 
If my understanding is correct then can we highlight this point in a comment here + also add a comment on top of ListPlugins of its response having both registered/non-registered thus all plugins.

Thinking how we can have  a similar plugin name validation on create time too. Does vault service has an API to expose the plugin name to UI for opting in/out on create? if yes, can we use that to validate here to make both flows similar?

Reply:
yes I can add comment, that would be very helpful

Vault service only has plugin APIs on the cluster level, so a cluster resource must exist on vault-service end first. (which is why we cant do the validation).

The options I see are:
1. Hard-coded validation. (personally not a fan , but can be ok if we aren't adding plugins frequently)
2. new vault-service API to serve plugins from feature flag for the LD rule of that organization.
3. What we are doing now where the terraform will fail when the plugin fails to add. (I believe the cluster will still be running and it should be doable for the user to update their terraform resource)
4. Making new `vault_plugin` resource that is not nested in `hcp_vault_cluster`, so it can fail independently of cluster creation and it requires an `hcp_vault_cluster` resource

Reply:
> 2. new vault-service API to serve plugins from feature flag for the LD rule of that organization.
I thought this already existed for HCP Portal but realizing may be portal is reading the values directly from FF?

> 3. What we are doing now where the terraform will fail when the plugin fails to add. (I believe the cluster will still be running and it should be doable for the user to update their terraform resource)
Hmm does this cause a concern in terraform state where cluster exists though the plugins given in input of create did not apply.

Reply:
I feel like 3 (current) is probably alright for now given that's the behavior on the UI - like you have to have created the cluster _first_ before you can add plugins, and if that flow fails, your cluster is still running.

> it should be doable for the user to update their terraform resource

Yup, users should be able to update the state using `terraform refresh` if it fails. (Not sure if you're aware, but if you wanted, you can test this and any real creates/updates in prod with your changes in local binary following [this guide](https://github.com/hashicorp/terraform-provider-hcp/blob/main/contributing/breaking-changes.md#how-to-test-existing-resources).)

I think long-term 2 or 4 would make more sense, 2 being an easier change and 4 being a more difficult one + maybe separating out other resources for consistency too.

Reply:
For 4 -> There was a FR in talks in another channel where the user was showing the desire to manage the state of configuration like observability, separate than cluster resource in terraform. So it may be that in some long term future we actually do 4 option for all major configurations. 

For now, agree that this flow seems fine.

Reply:
>> 2. new vault-service API to serve plugins from feature flag for the LD rule of that organization.

>I thought this already existed for HCP Portal but realizing may be portal is reading the values directly from FF?

The HCP portal API does use the FF, but the API also requires a valid cluster_id (which is why it can't be called before create).

going to stick with #3 for now.